### PR TITLE
create Division ID Governance OCDEP

### PR DESCRIPTION
**Pull request by djbridges**
A proposed governance model for OCD division IDs, as discussed in the Ballot Data Working Group

### CODE REVIEW

**Comment by kep** [comment]
These are valuable initial structures for building out a robust division-ID governance process. Thank you, Donny!

**Comment by afomi** [comment]
@djbridges - being explicit about the processes surrounding this data project is quite valuable. Thanks!

I'm not familiar with the .rst format, but think this is a useful draft. I'm curious about the implementation of the process. Given this process is encoded in this repo, I'm wondering how accountability will be assessed and enforced.

**Comment by fgregg** [comment]
HI @afomi. Could you expand a little bit on what you mean by "I'm wondering how accountability will be assessed and enforced." What are you thinking about?

**Comment by afomi** [comment]
I'm thinking that having a codified process is valuable in and of itself.
But, the execution of that process happens in people systems, instead of technical systems (ie: a compiler and runtime). 
So, I was just thinking about the implications of having this process codified inside the repo. Is it the best place for it? Would the wiki be better suited during draft time?

Overall, code in the repo represents a source of truth.
And while I really like the idea of codifying the people processes (the governance), it just strikes me as a separate, but related concern --- making me question its implications of being in code, in the repository. 

With that said, its a loosely held opinion of mine. I'm curious overall how governance of open projects is implemented and how it evolves.

**Comment by fgregg** [comment]
Hi @afomi, the first OCDEP describes the OCDEP process and is a great place to start: http://docs.opencivicdata.org/en/latest/proposals/0001.html  But, briefly, the OCDEP process has been working pretty well for the Open Civic Data community so far.

For a bit more context, the OCDEP process is heavily influence by the way that the Python community handles big changes: https://www.python.org/dev/peps/pep-0001/. This in turn is part of a tradition of self-governance on the internet that include the [RFC Process](https://en.wikipedia.org/wiki/Request_for_Comments) (which worked out the way that world web word work).

Hope that helps!

**Comment by afomi** [comment]
The extra context is extra helpful. Thanks @fgregg!

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: s/publicaly/publicly/

What do you mean by "open civic data" website? opencivicdata.org (which is controlled by @jamesturk) currently redirects to http://docs.opencivicdata.org/en/latest/

Do you mean docs.opencividata.org or something else?

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: Let's not name the particular technologies (i.e. Slack, Google Group) that we are using now, as that might change in five years.

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: How will the be advised? Throught a CONTRIBUTING.RST file or some other mechanism?

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: s/repsonsibilties/responsibilities/

What does it mean to "apply"?

Comment (fgregg) [reply proposals/drafts/ocdid_governance.rst]: Like send an e-mail to the mailing list?

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: Need some definition of a quorum probably.

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: Will every individual with a commit bit need to be approved be current committers?

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: How would revocation happen?

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: By whom. All committers will be reviewed by all other committers?

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: s/Accellerated/Accelerated/

Again, let's not name specific technology (Slack)

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: Even if the person is not a committer, but just a contributor?

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: This sentence got a little mangled.

File: proposals/drafts/ocdid_governance.rst
Comment (fgregg) [inline proposals/drafts/ocdid_governance.rst]: We'll these be automated checks.

**Review by fgregg — CHANGES_REQUESTED** [review]
Hi Donny,

Thanks for getting this started. Looks like a great start. Here's my initial feedback.

File: proposals/drafts/ocdid_governance.rst
Comment (jpmckinney) [reply proposals/drafts/ocdid_governance.rst]: This seems dangerous. I understand that timely review is a present issue, but once we implement the rest of the steps, it should be largely mitigated. (Introducing this raises the worst-case scenario of being bound to accept a PR that deletes all files.)

File: proposals/drafts/ocdid_governance.rst
Comment (jpmckinney) [inline proposals/drafts/ocdid_governance.rst]: I think we need there to be geographic segmentation of committers. A committer for Antarctica shouldn't have to participate in US votes/actions.

File: proposals/drafts/ocdid_governance.rst
Comment (jpmckinney) [reply proposals/drafts/ocdid_governance.rst]: GitHub operates on an individual level. It's unmanageable (and probably undesirable) to add an entire organization as committers (e.g. Google). If we are operating on an organizational basis, the organization should coordinate with committers which (small number of) people will be committers, and should notify committers when the individual leaves the organization… What's the (real) problems with individuals only? Bodies like the W3C have individuals on working groups representing organizations, but they are still individuals. The working groups don't have organizational membership.

**Review by jpmckinney — CHANGES_REQUESTED** [review]
Generally good! Please spell-check (many typos besides those @fgregg noted).

There should be a discussion of geographic segmentation. I'm not aware of any other non-US contributors, but the proposal doesn't take into account that user communities are largely segmented globally based on country of residence, and that this should be reflected in governance.

Less abstractly, I don't want an over-eager US committer making changes to Canada. I also don't want to have to participate in all the US approvals, etc. in order to maintain good standing. I have good processes in place for Canada, and would be responsive to Canadian PRs if there were any (I think there's only been one).

File: proposals/drafts/ocdid_governance.rst
Comment (jamesturk) [reply proposals/drafts/ocdid_governance.rst]: +1 to that, let's avoid 'organizations' - the one issue I see is that the delay in approval might hold an org up, but that could either be addressed through asking for an expedited decision or encouraging each org to have two committers

File: proposals/drafts/ocdid_governance.rst
Comment (jamesturk) [reply proposals/drafts/ocdid_governance.rst]: agree that this either needs to be fleshed out or omitted

File: proposals/drafts/ocdid_governance.rst
Comment (jamesturk) [reply proposals/drafts/ocdid_governance.rst]: agree.  I think updating this document w/ committers and their jurisdiction would be useful, perhaps ~3 tiers

- a few people might be considered 'global' and their participation could be requested in all votes (but we'd presumably at least informally waive the voting requirement?)
- most people would be scoped to a country
- any gov't officials would probably be scoped to their jurisdiction and thus not expected to participate in wider conversations

**Comment by jpmckinney** [comment]
FYI, some comments are appearing as "Show outdated" but only because typos were corrected - there are still substantive comments within those sections, so please expand those while reviewing.

**Comment by jamesturk** [comment]
Added some inline comments, agree w/ the need for geographic scope.  Also, where is the list of participants going to be published?  I'd be fine w/ using this doc, once accepted as the canonical source, or perhaps it'd be better to have this entered as two OCDEPs, one that is the this document and a second that is the current membership/governance & will be updated more frequently.

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: sgtm

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: Yes, I meant having them in the docs. We had also discussed creating a more friendly, splash-pagey introduction to OCDIDs for folks who are just discovering them (especially government folks) where they could be listed, but I see the docs being the place that gets canonically updated.

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: yep

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: "Apply" is maybe too heavy a word here. Going with "request" instead. And, yeah, I'm imagining they send an email/post in a slack channel/whatever we're doing in five years, then that request will get forwarded to the group of current contributors to see if there are any objections. Making that more explicit.

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: Sure

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: My concerns were mostly around giving organizations the flexibility to designate/manage their own committers for efficiency's sake, and also ensuring that an organization with a large number of members couldn't wield outsized voting power. I'm fine going individual only and letting community norms/processes be a check.

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: Adding in a detailed process mostly borrowed from http://docs.openvswitch.org/en/latest/internals/committer-grant-revocation/

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: heard

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: ack

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: @fgregg I'm thinking yes, as long as they're not a new contributor (which shouldn't be an issue, assuming the new committer is following the 'consult with a current committer' guideline in the COMMITTING.rst). I'm imagining the scenario where this most often would go into effect for a contributor would be for minor changes/corrections, as folks who would be making larger changes would probably either want to become or will be part of an organization that at least has a committer. If you think that's a bad assumption, or have strong feelings otherwise, happy to continue discussing.

Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: @jpmckinney I hear you and agree that it should be largely mitigated, but still think having an explicit backstop against things languishing in PR-purgatory is important. What if instead of straight auto-approve, a PR that hasn't been touched in 2 months gets pro-actively sent to the whole committer group for a "speak now or forever hold your peace" style review? It can still be approved by silence, if, say, it's about an esoteric thing that no current committers have knowledge of/an opinion on, but it ensures there are eyes on it to avoid a "delete everything" scenario?

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: Proofreading is fine for some. I like to live dangerously.

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: I would love for as many as possible to be, and think that there is a lot of opportunity for automated validation. I think there's always going to be a little bit of need for a human touch though, especially for things like determining whether a new flavor of special district is distinct enough to deserve its own type and the like. I also think there should still be more rigorous guidelines written before these automated checks get implemented.

File: proposals/drafts/ocdid_governance.rst
Comment (jpmckinney) [reply proposals/drafts/ocdid_governance.rst]: I'm not familiar with explicit backstops in open source development. Is there precedent for that? (preferably larger projects)

If literally no committer has an interest or opinion, I think that's to be solved by adding more knowledgeable committers – not by silently approving PRs. Divisions are political. Let's take fictional Kharun and South Kharun from the show Borgen. We get a PR that puts a division in the hierarchy of one country. But in fact, it's contested territory. I'd rather have someone with  political acuity chime in, than approve that change just because no present committer cares.

In the real world, if we got a PR for `ocd-division/country:cn/province:tw` for example (making Taiwan a province of China), it would at least make me think before merging. Whether we'd issue both that and `ocd-division/country:tw` to satisfy both views on the matter, I don't know, but I know it'd require some care, and shouldn't be automatic.

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: Completely agree on the need for geographic segmentation, and like the tiers. Also adding in a caveat about flexibility of process/governance at a country level.

File: proposals/drafts/ocdid_governance.rst
Comment (djbridges) [reply proposals/drafts/ocdid_governance.rst]: Those are both interesting cases that I agree would be tricky to handle, though even in the construction I mentioned above could still be prevented from getting automatically accepted by someone saying "I'm not sure what's going on here but maybe we should have a longer conversation/reach out to an expert." 

What I see as a more common way this would arise, though admittedly it may only happen in the US, is this: a researcher's gets on the phone with Elko, Nevada and discovers they have a separately elected [television control board](http://www.elkocountynv.net/departments/clerk/2012%20TV%20District%208%20of%209.pdf) that not only has it's own distinct geography, but it has [geographic subdistricts](https://www.dropbox.com/s/ow4c4u601a9n4gd/Screenshot%202017-11-01%2019.15.04.JPG?dl=0) from which people are elected. They create``ocd-division/country:us/state:nv/county:elko/television:e`` (or some such) and want to add it to the repo. Repeat any number of times for any number of weird districts, precinct name changes, etc. I completely get why someone wouldn't feel comfortable giving that an affirmative approval on face, but at some point it starts affecting how comprehensive tools using OCDIDs can be.

If there's a better way of handling that sort of use case that I'm missing, I'm all for it. Otherwise I think the sheer volume of those cases, in the US at least,  means the benefit of allowing them a graceful way to get in outweighs the potential for harm.

Comment (jpmckinney) [reply proposals/drafts/ocdid_governance.rst]: I think this is partly solved by having (geographically scoped) committers who are more lax with respect to performing verification and who are happy to merge things from people who appear to have good knowledge of a subject area. Basically, I still think this problem is solved at the level of people (who is deciding to merge), not at the level of process (when to auto-merge, etc.).

As we progress further down the hierarchy, the consequences of a bad merge become more and more limited. So, I'm fine with the governance of television control boards divisions being, "You say it exists? Sure! I'll merge."

We have some identifiers for countries for which no committer has sophisticated knowledge, and we have no identifiers for most countries. I think, for all such countries, we need an exemption from any sort of process-based auto-merge-like rule, and we also need an exemption for the top-level of the OCDID hierarchy (`ocd-division/country:`). Failure to adequately protect those with people-based governance can severely damage the project.

I'd also be in favor of scoping the auto-merge-like rule to specific geographies and/or specific division types (like `television`). I don't think the type `state` in the US should fall under the auto-merge-like rule – but I'll let US committers decide. For Canada, there's very low demand for PRs, so I'm not sure I need to implement such a rule at this time.

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst]: Shall we add a specific email group that a user is expected to use for making a contribution?

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst:35]: Just wanted to note that it is important for us to encourage user participation by making the adoption bar low for the technology to be used for group discussion/communication. Specifically if a user need to be on a specific platform that requires additional account creation, that may not be ideal to try to include more people who can participate on the project, that said, participating using an existing email address might be the minimum requirement.

**Comment by todrobbins** [comment]
This may be an uninformed question, but how would I become involved/join the Ballot Data Working Group?

Cheers!

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst:82]: This may be a high bar for many authoritative data sources (e.g. government), one of the common feedback we receive from state/county government is that they are particularly uncomfortable figuring out how github works. Will we offer additional help for creating pull requests on their behalf and/or integrate such steps as part of the contribution process?

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst:47]: Does this refer specifically to the earlier use case when i18n comitters are requested for voting regardless of the country? It is not 100% clear and/or maybe I'm missing some context.

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst:119]: Are we okay handling these cases via aliases? That way, it may not be an entire deviation but more of a variant due to naming conventions.

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst:117]: In case a government contributor is committer themselves, are they still required to work directly with an existing committer to prepare the pull request (step above)? Also, it may be reasonable to require that a second committer is not within the same government org for the approval?

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst:115]: Is this a requirement or optional if a government contributor doesn't feel comfortable handling a PR?

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst:121]: Shall we make this a requirement? Aliasing seems to be sufficient to handle naming conventions (and potentially other use cases) for OCD-IDs that are different and/or preferable by the government EC.

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst:123]: By conflict, do we mean differences between an existing OCD-ID and a government proposed OCD-ID? Wonder if that shall have been handled by aliases per the above step. Also, by marking official data in the repository, do we have any specific mechanism/approach in mind - e.g. 'sameAsNote'?

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst:130]: I'd make it explicit that this is more than a 'style' guide but more of a comprehensive 'how-to' guide. :)

File: proposals/drafts/ocdid_governance.rst
Comment (ccongchen) [inline proposals/drafts/ocdid_governance.rst:131]: Will there be reviews to check this, wonder over which time frame this participation bar applies.

**Comment by djbridges** [comment]
@todrobbins - I'll connect with you to give you some more info when we're getting back together.

@jpmckinney @jamesturk @fgregg  - Would love to get your feedback on how this iteration handles some of your earlier concerns! Thanks!

File: proposals/drafts/ocdid_governance.rst
Comment (jpmckinney) [inline proposals/drafts/ocdid_governance.rst:56]: Missing space after double backticks.

File: proposals/drafts/ocdid_governance.rst
Comment (jpmckinney) [inline proposals/drafts/ocdid_governance.rst:60]: If a country gets a large number of committers, it may be impractical to achieve a majority – and we may end up with the same situation that this OCDEP is trying to solve, which is quicker PR feedback and merging. Could we have "A majority, or #, of a country's `committers` (whichever is smaller)", where `#` is some fixed number?

File: proposals/drafts/ocdid_governance.rst
Comment (jpmckinney) [reply proposals/drafts/ocdid_governance.rst:131]: 60% may be a high bar. I'm not sure what 'actions' refers to, but if approving PRs is an action, then it may be hard to get there, if other committers are faster than you are.

**Review by jpmckinney — APPROVED** [review]
Looks good to me, but would want to get at least one other approval. My only concern is around the level of participation in votes/actions.

**Review by fgregg — APPROVED** [review]
Looks good, Donny. Thanks for your work on this.

**Comment by jpmckinney** [comment]
Also, I think @ccongchen's questions are good, but may not need to be answered in the OCDEP itself.

**Comment by jamesturk** [comment]
I'm in favor, I assume some aspects of this will be tweaked but think this gives us a good foundation to build upon.

**Pull request merged**
